### ISSUE TITLE
Question for CTC decoding

### ISSUE THREAD

**Issue by chengzhanzhan** [bgshih/crnn#10]
I don't clear that why the first position of time sequence always predicted as the first char of given string, just like:

![default](https://cloud.githubusercontent.com/assets/8745186/15536287/483de028-22a3-11e6-9688-120b8d8634eb.PNG)

I think that the first lable for most given sequence should be a blank space, such as

![imgres](https://cloud.githubusercontent.com/assets/8745186/15536598/98372458-22a4-11e6-9ee6-d3e008deb78a.png)

What do you think?

**Comment by bgshih** [comment]
@chengzhanzhan I have also observed the same phenomenon. Perhaps it's not a bug, but a strategy learned by LSTM. Notice that the CTC loss function does not care the position of individual letters, but only the final recognition results.

**Comment by chengzhanzhan** [comment]
@bgshih I test 38k images for this phenomenon. Because of the convolution operation, a section of continuous pixels are represented by a shorter interval, which may results in the first position including some information of the first character. 

Then I conduct experiments based on the 
**setting 1**: 3 * 256 * 32 (channel * width * height) image ==> convolution layers ==> 128 * 65* 1 ==> LSTM layers ==> CTC. Here, each column (of 65) represent several columns of origin image.

**setting 2**: 3 * 256 * 32 (channel * width * height) image ==> convolution layers ==> 128 * 225* 1 ==> LSTM layers ==> CTC to make sure that each step (of 225) represents each column of the image.

A example is as follow:

given image:
![121457](https://cloud.githubusercontent.com/assets/8745186/15595636/78f205ea-23f3-11e6-923e-72526dcf9771.jpg)

setting 1 result:
3-----------22----6----22-----b----------------------------------
shows that first position is character 3.

setting 2 result:
-----33------------------2----------------------6-------------------22-------------------------b---------------------------------------------------------------------------------------------------------------------------------
shows that first several positions is blank.

In CTC, each peak represents corresponding character, otherwise represent blank. Therefore, I think that the convolution strategy results in the phenomenon.

**Comment by ahmedmazari-dhatim** [comment]
Hi @bgshih @chengzhanzhan ,

How come that CTC loss doesn't care about the position of individual letter since each frame represent either a character or blank. So from my understating for each frame we have a logit which is transformed to probability which gives the letter or blank char in a given frame.

Sorry for the long comment but l don't understand the mapping between the output of RNN and  ground truth label . For instance : 
Input to CNN : image contains sequence "hello"

Hello => CNN => feature vector for each frame 

Input to RNN : 
 **A set of feature vectors representing each frame => B-LSTM => output logits  => CTC => Label = Hello"**


Do you mind explaining to me the process of transforming logits to characters ?
from this l understand that CTC predict the character for each frame.

I would be very grateful if you correct my understanding. Thank you

**Comment by githubharald** [comment]
@chengzhanzhan just want to let you know that I've observed the same phenomena with a different model, different framework (TensorFlow) and different input data (IAM). The last character is always predicted at the last time-step of the RNN output sequence. All other character-predictions are roughly aligned with the characters in the image. Look at the attached image, where the probability of the characters over time is plotted in the bottom-most graph: the "i", the "l"s and "t"s are aligned with the image while the final "e" is not.

![rnn_output](https://user-images.githubusercontent.com/15148095/41667567-94849586-74ad-11e8-910a-f027a96de01c.png)

**Comment by sheirving** [comment]
@githubharald 
Thankyou for your experimtent, do you figure out the reason caused this phenomena?

**Comment by githubharald** [comment]
no.

**Comment by JimmyJuan** [comment]
@githubharald Hello, can you tell me how to calculate the character-level probability in ctc-beam search? Is there a tutorial of implementing that in tensorflow?

thank you

**Comment by githubharald** [comment]
@Qhuangcn: see my repo CTCDecoder.

**Comment by luvwinnie** [comment]
does anyone research on ctc of this phenomenon, for my model it shows the same problem as this. for example a blank image input to this, it sometime decode something instead of blank only output...... This is a serious problem on OCR application.
Let say I have 4 fields have to recognition, and 1 field is optional to fill in on the document, it should be able to recognition as blank image as blank output. if not the recognition accuracy within a document will be affected
